ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactorUtil.java

```diff
@@ -295,4 +324,232 @@ public static LockRequest createLockRequest(HiveConf conf, CompactionInfo ci, lo
+        ci.type = type;
+        return requestCompaction(ci, runAs,hostName, ci.initiatorVersion, txnHandler);
+      }
+    } catch (InterruptedException e) {
+      //Handle InterruptedException separately so the compactionInfo won't be marked as failed.
+      LOG.info("Initiator pool is being shut down, task received interruption.");
+    } catch (Throwable ex) {
+      String errorMessage = "Caught exception while trying to determine if we should compact " + ci + ". Marking "
+              + "failed to avoid repeated failures, " + ex;
+      LOG.error(errorMessage);
+      ci.errorMessage = errorMessage;
+      if (metricsEnabled) {
+        Metrics.getOrCreateCounter(MetricsConstants.COMPACTION_INITIATOR_FAILURE_COUNTER).inc();
+      }
+      txnHandler.markFailed(ci);
+    }
+    return null;
+  }
+
+  public static CompactionResponse initiateCompactionForPartition(Table table, Partition partition,
+                                                           CompactionRequest compactionRequest,String hostName, TxnStore txnHandler, HiveConf conf) throws MetaException {
+    ValidTxnList validTxnList = TxnCommonUtils.createValidReadTxnList(txnHandler.getOpenTxns(), 0);
+    conf.set(ValidTxnList.VALID_TXNS_KEY, validTxnList.writeToString());
+    CompactionResponse compactionResponse;
+    CompactionInfo compactionInfo =
+            new CompactionInfo(table.getDbName(), table.getTableName(), compactionRequest.getPartitionname(),
+                    compactionRequest.getType());
+    compactionInfo.initiatorId = compactionRequest.getInitiatorId();
+    compactionInfo.orderByClause = compactionRequest.getOrderByClause();
+    compactionInfo.initiatorVersion = compactionRequest.getInitiatorVersion();
+    if (compactionRequest.getNumberOfBuckets() > 0) {
+      compactionInfo.numberOfBuckets = compactionRequest.getNumberOfBuckets();
+    }
+    compactionInfo.poolName = compactionRequest.getPoolName();
+    try {
+      StorageDescriptor sd = CompactorUtil.resolveStorageDescriptor(table, partition);
+      String runAs = TxnUtils.findUserToRunAs(sd.getLocation(), table, conf);
+      LOG.info("Checking to see if we should compact partition {} of table {}.{}",
+              compactionInfo.partName, table.getDbName(), table.getTableName());
+      compactionResponse = CompactorUtil.scheduleCompactionIfRequired(compactionInfo, table, partition, runAs, false, hostName, txnHandler, conf);
+    } catch (IOException | InterruptedException | MetaException e) {
+      LOG.error("Error occurred while Checking if we should compact partition {} of table {}.{} Exception: {}",
+              compactionInfo.partName, table.getDbName(), table.getTableName(), e.getMessage());
+      throw new RuntimeException(e);
+    }
+    return compactionResponse;
+  }
+
+  public static CompactionResponse initiateCompactionForTable(CompactionRequest request, TxnStore txnHandler) throws MetaException {
+    return txnHandler.compact(request);
```

Comment:
In the previous version of the code before launching the compaction the code was setting the `VALID_TXNS_KEY`
```
ValidTxnList validTxnList = TxnCommonUtils.createValidReadTxnList(txnHandler.getOpenTxns(), 0);
    conf.set(ValidTxnList.VALID_TXNS_KEY, validTxnList.writeToString());
```
Is it intentional that now it is not done?

Most of the other comments are nitpicks but this is important to clarify.

Reply:
The code before this refactoring is not failing so I am trying to understand what is causing this change in behavior. I don't want to let changes in behavior break in the code base accidentally. Can you please elaborate why it was necessary to do it before and it shouldn't be done now?

Other than that if the method is simply a call to `txnHandler.compact(request)` then it is not really necessary to wrap it in a utility. Using `txnHandler.compact(request)` directly is shorter and more explicit.

Reply:
Sorry for not clarifying this part.

Before making changes as part of this pull request, 
`ValidTxnList validTxnList = TxnCommonUtils.createValidReadTxnList(txnHandler.getOpenTxns(), 0);
    conf.set(ValidTxnList.VALID_TXNS_KEY, validTxnList.writeToString());`

is in the method  (initiateCompactionForTable|https://github.com/apache/hive/blob/a348e5d86adc611e4f72d386815a21a96b667ab4/ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/InitiatorBase.java#L96) of the InitiatorBase class 

So earlier also, this is not part of AlterTableCompactOperation.

So functionally there is no change now as it is moved to CompactorUtil.initiateCompactionForPartition

Earlier initialization of TxnHandler is taken care in (MetaStoreCompactorThread|https://github.com/apache/hive/blob/a4d4b9bf3cbd04b9bad13068a13fff22e6ad9e70/ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/MetaStoreCompactorThread.java#L60) due to inheritance. Now we need to initialize it in AlterTableCompactOperation.

After setting `ValidTxnList.VALID_TXNS_KEY` in conf of AlterTableCompactOperation, test cases are failing with error,
`Need to record valid WriteID list but there is no valid TxnID list `

That's why moved this into the original method CompactorUtil.initiateCompactionForPartition

Hope this is clear now. Thank you

Reply:
Actually, I see that in the [old code](https://github.com/apache/hive/blob/764c1a2fd2e645d8fa49ad6fbabf483342ccaaf7/ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/InitiatorBase.java#L101) before calling `txnHandler.compact(request)` we used to do [conf.set](https://github.com/apache/hive/blob/764c1a2fd2e645d8fa49ad6fbabf483342ccaaf7/ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/InitiatorBase.java#L97) and now we don't for this code path.

Reply:
Yes. 

setting conf with `ValidTxnList.VALID_TXNS_KEY` is only needed for queries like alter table compact on partitioned tables, which needs to check & identify eligible partitions to get compacted

Earlier we are doing that for `initiateCompactionForMultiplePartitions` as we are calling it within the method `initiateCompactionForTable`. But setting conf with `ValidTxnList.VALID_TXNS_KEY` is not needed for `txnHandler.compact(request)`.  That's why we are not doing it in the newer version where we have separated methods for table and partitions

> The new CompactorUtil.initiateCompactionForPartition method, which accepts context.getConf, modifies the original configuration and not the copy.
> 
> The change above may be related to the test failures that happen when setting the VALID_TXNS_KEY in various places.

Yes, earlier InitiatorBase is a separate Class/Thread runs as a service and so it has it's own conf due to inheritance.
But now we have switched to using static methods, need to use the same config from the AlterTableCompactOperation class. 

Please advise if any changes can be made w.r.t new config object creation in the static method `initiateCompactionForPartition`

Reply:
The main question is why it is not needed.

I checked the code for `txnHandler.compact(request)` in more detail and now I can see why it is not necessary to set  `VALID_TXNS_KEY` in this case. The ValidTxnList is gonna be retrieved later on by [GetValidWriteIdsFunction](https://github.com/apache/hive/blob/ffb1165f59defa66b31b4fd9cb6367b71050071b/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/jdbc/functions/GetValidWriteIdsFunction.java#L64).
Even if it was set in the previous version of the code it seems that it didn't have really any effect on `txnHandler.compact(request)` so it was redundant in the first place.

Regarding the config modification, I will start (or comment) in a separate thread to avoid mixing up the discussions.

I will now mark this as resolved.